File: Design-Patterns-Using-CSharp/Mediator/MediatorPattern2.cs
using System;
using System.Collections.Generic;

interface IMediator
{
    void AddColleague(Colleague colleague);
    void SendMessage(string message, string colleagueId);
}

class ConcreteMediator : IMediator
{
    private Dictionary<string, Colleague> colleagues = new Dictionary<string, Colleague>();

    public void AddColleague(Colleague colleague)
    {
        colleagues[colleague.GetId()] = colleague;
    }

    public void SendMessage(string message, string colleagueId)
    {
        Console.WriteLine($"Mediator pass Message : {message}");
        colleagues[colleagueId].Receive(message);
    }
}

abstract class Colleague
{
    protected string id;
    protected IMediator mediator;

    public Colleague(string id, IMediator mediator)
    {
        this.id = id;
        this.mediator = mediator;
    }

    public abstract void Send(string message, string to);

    public abstract void Receive(string message);

    public string GetId()
    {
        return id;
    }
}

class ConcreteColleague : Colleague
{
    public ConcreteColleague(string id, IMediator mediator) : base(id, mediator) { }

    public override void Send(string message, string to)
    {
        Console.WriteLine($"{id} Sent Message : {message}");
        mediator.SendMessage(message, to);
    }

    public override void Receive(string message)
    {
        Console.WriteLine($"{id} Received Message {message}");
    }
}

// Client code

public class MediatorPattern2
{
    public static void Main(string[] args)
    {
        ConcreteMediator mediator = new ConcreteMediator();
        ConcreteColleague first = new ConcreteColleague("First", mediator);
        mediator.AddColleague(first);

        ConcreteColleague second = new ConcreteColleague("Second", mediator);
        mediator.AddColleague(second);

        first.Send("Hello, World!", "Second");
        second.Send("Hi, World!", "First");
    }
}

/*
First Sent Message : Hello, World!
Mediator pass Message : Hello, World!
Second Received Message Hello, World!
Second Sent Message : Hi, World!
Mediator pass Message : Hi, World!
First Received Message Hi, World!
*/

File: Design-Patterns-Using-Go/Mediator/MediatorPattern2.go
package main

import "fmt"

// Mediator interface
type Mediator interface {
	addColleague(colleague Colleague)
	sendMessage(message string, colleagueID string)
}

// ConcreteMediator struct
type ConcreteMediator struct {
	colleagues map[string]Colleague
}

func (m *ConcreteMediator) addColleague(colleague Colleague) {
	m.colleagues[colleague.getID()] = colleague
}

func (m *ConcreteMediator) sendMessage(message string, colleagueID string) {
	fmt.Printf("Mediator pass Message: %s\n", message)
	m.colleagues[colleagueID].receive(message)
}

// Colleague interface
type Colleague interface {
	getID() string
	send(message string, to string)
	receive(message string)
}

// ConcreteColleague struct
type ConcreteColleague struct {
	id       string
	mediator Mediator
}

func (c *ConcreteColleague) getID() string {
	return c.id
}

func (c *ConcreteColleague) send(message string, to string) {
	fmt.Printf("%s Sent Message: %s\n", c.id, message)
	c.mediator.sendMessage(message, to)
}

func (c *ConcreteColleague) receive(message string) {
	fmt.Printf("%s Received Message %s\n", c.id, message)
}

// Client code
func main() {
	mediator := &ConcreteMediator{
		colleagues: make(map[string]Colleague),
	}

	first := &ConcreteColleague{
		id:       "First",
		mediator: mediator,
	}
	mediator.addColleague(first)

	second := &ConcreteColleague{
		id:       "Second",
		mediator: mediator,
	}
	mediator.addColleague(second)

	first.send("Hello, World!", "Second")
	second.send("Hi, World!", "First")
}
